txnkv/transaction/txn.go

```diff
@@ -526,6 +559,12 @@ func (txn *KVTxn) LockKeysWithWaitTime(ctx context.Context, lockWaitTime int64,
 // LockKeys tries to lock the entries with the keys in KV store.
 // lockCtx is the context for lock, lockCtx.lockWaitTime in ms
 func (txn *KVTxn) LockKeys(ctx context.Context, lockCtx *tikv.LockCtx, keysInput ...[]byte) error {
+	if txn.interceptor != nil {
+		// User has called txn.SetInterceptor() to explicitly set an interceptor, we
+		// need to bind it to ctx so that the internal client can perceive and execute
+		// it before initiating an RPC request.
+		ctx = tikvrpc.SetInterceptorIntoCtx(ctx, txn.interceptor)
+	}
```

Comment:
Just an idea, needs to research: in L590 client-go let the caller to insert values into the ctx:

```go
ctxValue := ctx.Value(util.LockKeysDetailCtxKey)
if ctxValue != nil {
	lockKeysDetail := ctxValue.(**util.LockKeysDetails)
	*lockKeysDetail = lockCtx.Stats
}
```

Example caller in TiDB:

```go
ctx = context.WithValue(ctx, tikvutil.LockKeysDetailCtxKey, &lockKeyStats)
err = txn.LockKeys(tikvutil.SetSessionID(ctx, se.GetSessionVars().ConnectionID), lockCtx, keys...)
```

Is it possible for us to just follow this pattern without introducing `txn.interceptor`? How complex will it becomes for TiDB?

My starting point is that, seems that when we should replace the `ctx` to attach the `txn.interceptor` is pretty tricky. It is very easy to be missed when a new RPC wrapper is added by other people. It would be nice if we don't use it.

Reply:
> Is it possible for us to just follow this pattern without introducing txn.interceptor? How complex will it becomes for TiDB?

I thought the same at the beginning, but the biggest problem is that not all client-go interfaces provide the `ctx` parameter, so there is no way to use context uniformly.

For example, the interface to iterate on the snapshot is:
```go
Iter(k []byte, upperBound []byte) (Iterator, error)
```

The internal Scanner implementation is:
```go
func (s *Scanner) Next() error {
	bo := retry.NewBackofferWithVars(context.WithValue(context.Background(), retry.TxnStartKey, s.snapshot.version), scannerNextMaxBackoff, s.snapshot.vars)
	if !s.valid {
		return errors.New("scanner iterator is invalid")
	}
```

The problem is that `context.Context` is constructed inside client-go.

Reply:
How about introducing `ctx` for them (if the change at TiDB side is not big then it should be great)? It seems to be bad when `Next()` is not controllable via `ctx`. There might be many changes at TiDB side. If so, please abandon this direction.

Reply:
It seems to be a very heavy work : (   The general interface in package `kv` has been defined like this:

```go
// Retriever is the interface wraps the basic Get and Seek methods.
type Retriever interface {
	Getter
	// Iter creates an Iterator positioned on the first entry that k <= entry's key.
	// If such entry is not found, it returns an invalid Iterator with no error.
	// It yields only keys that < upperBound. If upperBound is nil, it means the upperBound is unbounded.
	// The Iterator must be Closed after use.
	Iter(k Key, upperBound Key) (Iterator, error)

	// IterReverse creates a reversed Iterator positioned on the first entry which key is less than k.
	// The returned iterator will iterate from greater key to smaller key.
	// If k is nil, the returned iterator will be positioned at the last key.
	// TODO: Add lower bound limit
	IterReverse(k Key) (Iterator, error)
}
```

```go
// Iterator is the interface for a iterator on KV store.
type Iterator interface {
	Valid() bool
	Key() Key
	Value() []byte
	Next() error
	Close()
}
```